### PR TITLE
Fix NullRef in SemanticClassificationViewTaggerProvider

### PR DESCRIPTION
**Pull request by dpoeschl**
@balajikris @jasonmalinowski @rchande @KevinH-MS 
@dotnet/roslyn-ide

### CODE REVIEW

**Comment by KevinH-MS** [comment]
FYI @tannergooding (we should try to insert a build with this change at earliest convenience)

**Comment by dpoeschl** [comment]
This null check existed until a refactoring happened back in September that accidentally removed it.,

**Comment by Pilchie** [comment]
:+1:

**Comment by jasonmalinowski** [comment]
I find your lack of tests disturbing.

File: src/EditorFeatures/Core/Implementation/Classification/SemanticClassificationViewTaggerProvider.cs [hunk 1]
```diff
@@ -92,9 +92,12 @@ protected override Task ProduceTagsAsync(TaggerContext<IClassificationTag> conte
             Debug.Assert(context.SpansToTag.IsSingle());
 
             var spanToTag = context.SpansToTag.Single();
-            var document = spanToTag.Document;
+            if (spanToTag.Document == null)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Classification/SemanticClassificationViewTaggerProvider.cs]: I would far prefer that we filter out documentless spans in GetSpansToTag.

**Comment by CyrusNajmabadi** [comment]
:-1:

**Comment by davkean** [comment]
Mono crashed

retest prtest/lin/dbg/unit32 please

**Comment by dpoeschl** [comment]
@CyrusNajmabadi Fixed. Please re-review.

**Comment by dpoeschl** [comment]
retest vsi please

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs [hunk 2]
```diff
@@ -306,25 +306,30 @@ private List<DocumentSnapshotSpan> GetSpansAndDocumentsToTag()
...
+                Document document = null;
+                if (!snapshotToDocumentMap.TryGetValue(span.Snapshot, out document))
+                {
+                    CheckSnapshot(span.Snapshot);
+
+                    document = span.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+                    snapshotToDocumentMap[span.Snapshot] = document;
+                }
+
+                // document can be null if the buffer the given span is part of is not part of our workspace.
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs]: Update comment to mention that we will filter this out in the above query.

**Comment by CyrusNajmabadi** [comment]
:+1:

**Comment by Pilchie** [comment]
:+1:

**Comment by dpoeschl** [comment]
That's a lot of timeouts!

**Comment by dpoeschl** [comment]
ETA isn't reported as a timeout, but the log looks like it timed out. It's also missing it's artifacts. :cry:

**Comment by dpoeschl** [comment]
@dotnet-bot retest prtest/win/dbg/eta please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn_prtest_win_dbg_eta/2653/
// Retest reason: Builds were hanging

**Comment by dpoeschl** [comment]
@dotnet-bot retest prtest/win/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/2196/
// Retest reason: Builds were hanging

**Comment by dpoeschl** [comment]
@dotnet-bot retest prtest/win/vsi/p3 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn_prtest_win_vsi_p3/124/
// Retest reason: Builds were hanging

**Comment by dpoeschl** [comment]
@dotnet-bot retest prtest/win/vsi/p2 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn_prtest_win_vsi_p2/127/
// Retest reason: Builds were hanging

**Comment by dpoeschl** [comment]
@dotnet-bot retest prtest/win/vsi/p0 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn_prtest_win_vsi_p0/123/
// Retest reason: Builds were hanging

**Comment by dpoeschl** [comment]
@dotnet-bot retest prtest/win/vsi/p1 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn_prtest_win_vsi_p1/126/
// Retest reason: Builds were hanging

**Comment by dpoeschl** [comment]
@dotnet-bot retest prtest/win/dbg/unit64 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit64/2175/
// Retest reason: Builds were hanging

**Comment by dpoeschl** [comment]
@dotnet-bot retest prtest/win/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/2261/
// Retest reason: Timeout cloning repo from GitHub. Retry the job.

**Comment by dpoeschl** [comment]
p3: CSharpReferenceHighlighting hung, testing locally now.

**Comment by dpoeschl** [comment]
It's actually broken. Investigating.

**Comment by dpoeschl** [comment]
retest vsi please

**Comment by dpoeschl** [comment]
(If tests pass, I'll squash commits)

**Comment by dpoeschl** [comment]
Roslyn.Services.Editor.CSharp.UnitTests.dll unit tests hung in prtest/win/dbg/unit*

**Comment by dpoeschl** [comment]
retest ccc please
// Does this do anything?

**Comment by dpoeschl** [comment]
retest vsi please

**Comment by tannergooding** [comment]
@dpoeschl, Can we either fixup and merge this PR or close it?

**Comment by jaredpar** [comment]
We apologize, but we are closing this PR due to code drift. We regret letting this PR go so long without attention, but at this point the code base has changed too much for us to revisit older PRs. Going forward, we will manage PRs better under an SLA currently under draft in issue #26266 – we encourage you to add comments to that draft as you see fit.   Although that SLA is still in draft form, we nevertheless want to move forward with the identification of older PRs to give contributors as much early notice as possible to review and update them.

If you are interested in pursuing this PR, please reset it against the head of master and we will reopen it. Thank you!

**Pull request closed without merging**